src/System.IO.Pipelines.Networking.Tls/Internal/OpenSsl/CustomBio.cs

```diff
@@ -0,0 +1,253 @@
+            public extern static void BIO_set_flags(ref bio_st bio, BioFlags flags);
+        }
+
+        private static class UnixLib
+        {
+            public const string CryptoDll = "libcrypto.so";
+            [DllImport(CryptoDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static void BIO_set_flags(ref bio_st bio, BioFlags flags);
+        }
+        private static class OsxLib
+        {
+            public const string CryptoDll = "libcrypto.1.0.0.dylib";
+            [DllImport(CryptoDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static void BIO_set_flags(ref bio_st bio, BioFlags flags);
+        }
+
+        private static void BIO_set_flags(ref bio_st bio, BioFlags flags)
+        {
+            if (Interop.IsWindows)
+            {
+                WindowsLib.BIO_set_flags(ref bio, flags);
+            }
+            else if (Interop.IsOsx)
+            {
+                OsxLib.BIO_set_flags(ref bio, flags);
+            }
+            else
+            {
+                UnixLib.BIO_set_flags(ref bio, flags);
+            }
+        }
+
+        private static int FreeBio(ref bio_st bio)
+        {
+            if (bio.init == 1 && bio.shutdown == 1)
+            {
+                bio.ptr = null;
+                return 1;
+            }
+            return -1;
+        }
+
+        private static int CreateBio(ref bio_st bio)
+        {
+            bio.shutdown = 1;
+            bio.init = 1;
+            bio.num = -1;
+            return 1;
+        }
+
+        private static int WriteBio(ref bio_st bio, void* buff, int numberOfBytes)
+        {
+            var bufferHandle = GCHandle.FromIntPtr((IntPtr)bio.ptr);
+            var bufferPipeline = (IPipelineWriter)bufferHandle.Target;
+            int numberOfBytesRemaing = numberOfBytes;
+            var buffer = bufferPipeline.Alloc();
+            while (numberOfBytesRemaing > 0)
+            {
+                int sizeToWrite = Math.Min(MaxBlockSize, numberOfBytesRemaing);
+                buffer.Ensure(sizeToWrite);
+                var writeSpan = new Span<byte>(buff, sizeToWrite);
+                writeSpan.CopyTo(buffer.Memory.Span);
+                buffer.Advance(sizeToWrite);
+                numberOfBytesRemaing -= sizeToWrite;
+            }
+            buffer.Commit();
+            bio.num_write = (uint)numberOfBytes;
+            BIO_set_flags(ref bio, BioFlags.BIO_FLAGS_WRITE | BioFlags.BIO_FLAGS_SHOULD_RETRY);
+            return numberOfBytes;
+        }
+
+        public static int NumberOfWrittenBytes(InteropBio.BioHandle handle)
+        {
+            var bio = (bio_st*)handle.Handle;
+            var bytes = (int)bio[0].num_write;
+            bio[0].num_write = 0;
+            return bytes;
+        }
+
+        private static int ReadBio(ref bio_st bio, void* buff, int numberOfBytes)
+        {
+            var buffer = Unsafe.Read<ReadableBuffer>(bio.ptr);
```

Comment:
This might have changed, but last I looked (before the move to corefxlabs) and after long discussions between Ben and I, readablebuffer is stack only. Writable isn't and is pinned correctly and a GCHandle is stashed instead or a pointer to the stack struct like the readable